### DISCUSSION TITLE
Can I use android.widget.ImageButton in Toga？

### DISCUSSION THREAD

**Discussion by wanglifeng95** [beeware/toga#2087] [Q&A] [answered]
Toga can directly use Android native APIs directly. Now that Toga does not have an ImageButton widget, how can I add an android.widget.ImageButton to Toga.Box?

**Comment by freakboy3742** [comment, marked as answer]
I can't point you at a simple worked example, but: if you have an instance of `toga.Box()` called `mybox`, `mybox._impl.native` is an instance of `android.widget.RelativeLayout`, and `mybox.container.native_content` is another instance of `android.widget.RelativeLayout` that represents the top-level container for the window being laid out. You can then use [Chaquopy's APIs](https://chaquo.com/chaquopy/doc/current/python.html) to wrap the Java classes needed to construct an `android.widget.ImageButton`.

A normal Toga widget would be added to `mybox.container.native_content` and then have positioning parameters set as part of Toga's layout algorithm; if you were to add a widget to the `._impl.native` of a box, that widget should be inserted relative to the position of that box.

Alternatively, you could define a custom "ImageButton" widget, but only provide an Android implementation. I'd guess that 90% of the implementation would be identical to Toga's Button - the only real difference would be the parts that actually construct the Button, and the parts that direct the factory to construct the Android implementation.

**Reply by wanglifeng95** [reply]
Thank you very much for your reply, which gave me some ideas. I will conduct a specific test tomorrow

**Reply by mhsmith** [reply]
Note that the `native_content` attribute was added in the current development version of Toga, and doesn't exist on Toga 0.3.1 and earlier.

So I recommend you try adding a custom widget class instead. This will be more likely to keep on working with future versions of Toga, and will allow you to position your widget using Toga's own layout system.

**Reply by wanglifeng95** [reply]
Thank you both very much for your help. I know how to use Android's native APIs, and I have finally implemented clickable images on Android. 
Another question, I want to do something similar on Windows. How does Toga use Windows's native APIs?

**Reply by mhsmith** [reply]
On Windows it uses [Python.NET](https://pythonnet.github.io/), and you can find the relevant code in toga/winforms.

**Reply by wanglifeng95** [reply]
Okay, thank you very much. I'll find a time to study it carefully

**Reply by wanglifeng95** [reply]
I have uniformly implemented image buttons, image events, and pop-up menus for Android and Windows. Next, I want to do similar things on Linux systems such as Ubuntu, how does Toga use its native APIs?

**Reply by freakboy3742** [reply]
Toga uses:
* [PyGObject](https://pygobject.readthedocs.io) on Linux to talk to GTK APIs.
* [Rubicon ObjC](https://rubicon-objc.readthedocs.io/en/stable/) on iOS to talk to UIKit;
* [Rubicon ObjC](https://rubicon-objc.readthedocs.io/en/stable/) on macOS to talk to AppKit; and
* [Python.net](https://github.com/pythonnet/pythonnet) on Windows to talk to Winforms .NET APIs.

On each platform, the `widget._impl.native` attribute is the native representation of a widget, using whatever platform library is appropriate.

**Reply by wanglifeng95** [reply]
Okay, thank you very much for your guidance

**Comment by wanglifeng95** [comment]
您好，我已收到来信，会尽快给予你回复。
